File: ios/Classes/TargetingOptionsMethodHandler.swift
//
//  CASMethodHandler.swift
//  Runner
//
//  Created by Dmytro Uzhva on 13.08.2024.
//

import CleverAdsSolutions
import Flutter

class TargetingOptionsMethodHandler: MethodHandler {
    func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        switch call.method {
        case "getGender": getGender(result: result)
        case "setGender": setGender(call: call, result: result)
        case "getAge": getAge(result: result)
        case "setAge": setAge(call: call, result: result)
        case "getLocationLatitude": getLocationLatitude(result: result)
        case "setLocationLatitude": setLocationLatitude(call: call, result: result)
        case "getLocationLongitude": getLocationLongitude(result: result)
        case "setLocationLongitude": setLocationLongitude(call: call, result: result)
        case "isLocationCollectionEnabled": isLocationCollectionEnabled(result: result)
        case "setLocationCollectionEnabled": setLocationCollectionEnabled(call: call, result: result)
        case "getKeywords": getKeywords(result: result)
        case "setKeywords": setKeywords(call: call, result: result)
        case "getContentUrl": getContentUrl(result: result)
        case "setContentUrl": setContentUrl(call: call, result: result)
        default: return false
        }
        return true
    }

    private func getGender(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.gender)
    }

    private func setGender(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "gender", call: call, result: result) { gender in
            CAS.targetingOptions.gender = gender
        }
    }

    private func getAge(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.age)
    }

    private func setAge(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "age", call: call, result: result) { age in
            CAS.targetingOptions.age = age
        }
    }

    private func getLocationLatitude(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.locationLatitude)
    }

    private func setLocationLatitude(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "latitude", call: call, result: result) { latitude in
            CAS.targetingOptions.locationLatitude = latitude
        }
    }

    private func getLocationLongitude(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.locationLongitude)
    }

    private func setLocationLongitude(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "longitude", call: call, result: result) { longitude in
            CAS.targetingOptions.locationLongitude = longitude
        }
    }

    private func isLocationCollectionEnabled(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.locationCollectionEnabled)
    }

    private func setLocationCollectionEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "isEnabled", call: call, result: result) { isEnabled in
            CAS.targetingOptions.locationCollectionEnabled = isEnabled
        }
    }

    private func getKeywords(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.keywords)
    }

    private func setKeywords(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "keywords", call: call, result: result) { keywords in
            CAS.targetingOptions.keywords = keywords
        }
    }

    private func getContentUrl(result: @escaping FlutterResult) {
        result(CAS.targetingOptions.contentUrl)
    }

    private func setContentUrl(call: FlutterMethodCall, result: @escaping FlutterResult) {
        tryGetArgSetValue(name: "contentUrl", call: call, result: result) { contentUrl in
            CAS.targetingOptions.contentUrl = contentUrl
        }
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/TargetingOptionsMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge

import android.location.Location
import com.cleveradssolutions.plugin.flutter.bridge.base.MethodHandler
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import com.cleversolutions.ads.android.CAS
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/targeting_options"

class TargetingOptionsMethodHandler(
    binding: FlutterPluginBinding
) : MethodHandler(binding, CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getGender" -> getGender(result)
            "setGender" -> setGender(call, result)
            "getAge" -> getAge(result)
            "setAge" -> setAge(call, result)
            "getLocationLatitude" -> getLocationLatitude(result)
            "setLocationLatitude" -> setLocationLatitude(call, result)
            "getLocationLongitude" -> getLocationLongitude(result)
            "setLocationLongitude" -> setLocationLongitude(call, result)
            "isLocationCollectionEnabled" -> isLocationCollectionEnabled(result)
            "setLocationCollectionEnabled" -> setLocationCollectionEnabled(call, result)
            "getKeywords" -> getKeywords(result)
            "setKeywords" -> setKeywords(call, result)
            "getContentUrl" -> getContentUrl(result)
            "setContentUrl" -> setContentUrl(call, result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getGender(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.gender)
    }

    private fun setGender(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("gender", result) {
            CAS.targetingOptions.gender = it
        }
    }

    private fun getAge(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.age)
    }

    private fun setAge(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("age", result) {
            CAS.targetingOptions.age = it
        }
    }

    private fun getLocationLatitude(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.location?.latitude)
    }

    private fun setLocationLatitude(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Double>("latitude", result) {
            val location = CAS.targetingOptions.location ?: Location("").also {
                CAS.targetingOptions.location = it
            }
            location.latitude = it
        }
    }

    private fun getLocationLongitude(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.location?.longitude)
    }

    private fun setLocationLongitude(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Double>("longitude", result) {
            val location = CAS.targetingOptions.location ?: Location("").also {
                CAS.targetingOptions.location = it
            }
            location.longitude = it
        }
    }

    private fun isLocationCollectionEnabled(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.locationCollectionEnabled)
    }

    private fun setLocationCollectionEnabled(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            CAS.targetingOptions.locationCollectionEnabled = it
        }
    }

    private fun getKeywords(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.keywords?.toList())
    }

    private fun setKeywords(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<List<String>?>("keywords", result) {
            CAS.targetingOptions.keywords = it?.toSet()
        }
    }

    private fun getContentUrl(result: MethodChannel.Result) {
        result.success(CAS.targetingOptions.contentUrl)
    }

    private fun setContentUrl(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String?>("contentUrl", result) {
            CAS.targetingOptions.contentUrl = it
        }
    }

}